### PR TITLE
[V3 Core] Graceful shutdown when SIGTERM is received

### PR DESCRIPTION
**Pull request by laggron42**
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes

When the bot receives SIGTERM, it will close the connexion (with `bot.shutdown`) correctly. Without this handler, the process is killed and the gateway is stuck, with the bot still connected for a while. Using `sudo systemctl stop red.service <instance>` is now safe to correctly shutdown the bot.

What it looks like in the logs:
```
laggron@discordbots:~/RedDB-fork$ sudo systemctl stop redfork-discordbot.service
laggron@discordbots:~/RedDB-fork$ systemctl status redfork-discordbot.service
● redfork-discordbot.service - Red-DiscordBotV3:Red
   Loaded: loaded (/etc/systemd/system/redfork-discordbot.service; disabled; vendor preset: enabled)
   Active: inactive (dead)

[...]
Nov 03 11:50:28 discordbots pipenv[12401]: [03/11/2018 11:50] INFO __main__ sigterm_handler 115: SIGTERM received. Quitting...
Nov 03 11:50:28 discordbots pipenv[12401]: [03/11/2018 11:50] INFO manager shutdown_lavalink_server 115: Shutting down lavalink server.
```

### CODE REVIEW

**Comment by laggron42** [comment]
Note: this was only tested on Ubutnu 18.10 with systemdaemon, it should work the same with upstart, but I don't know how Windows handle this, so this might need testing on this OS (I don't have Red installed on Win)

File: docs/autostart_systemd.rst [hunk 1]
```diff
@@ -48,6 +49,14 @@ To set the bot to start on boot, you must enable the service, again adding the i
 
 :code:`sudo systemctl enable red@instancename`
 
+If you need to shutdown the bot, you can use the ``[p]shutdown`` command or
+you can stop the service, still by adding the instance name after the **@**:
```
Comment (Tobotimus) [inline docs/autostart_systemd.rst]: Grammar is slightly off here, maybe just say

> ...or you can stop the service with the following command:

File: redbot/__main__.py [hunk 2]
```diff
@@ -139,6 +145,7 @@ def main():
     if cli_flags.dev:
         red.add_cog(Dev())
     loop = asyncio.get_event_loop()
+    loop.add_signal_handler(SIGTERM, lambda: asyncio.ensure_future(sigterm_handler(red, log)))
```
Comment (Tobotimus) [inline redbot/__main__.py]: This method is not implemented on non-unix systems, so this needs to be guarded under an `if os.name == "posix"` condition.

**Review by Tobotimus — CHANGES_REQUESTED** [review]
Thanks for adding this. Just need a few changes and it should be good to go.

**Review by Tobotimus — APPROVED** [review]

**Pull request merged**